[16.0] endpoint_route_handler: migration to v16

simahawk opened the pull request:
Includes #3

simahawk committed:
Add endpoint_route_handler

simahawk committed:
endpoint_route_handler: reduce log noise

OCA-git-bot committed:
endpoint_route_handler 14.0.1.0.1

simahawk committed:
endpoint_route_handler: fix archive/unarchive

When an endpoint is archived it must be dropped.
When it's unarchive it must be restored.

simahawk committed:
endpoint_route_handler: fix multi env handling

Routing maps are generated **per env**
which means that every new env will have its own routing map
attached to `ir.http` registry class.

This is not desired (as per core Odoo comment)
but it's like this today :/

Hence, before this change, the routing map could be mis-aligned
across different envs leading to random responses for custom endpoints.

This refactoring simplifies a lot the handling of the rules
leaving to std `_generate_routing_rules` the duty to yield rules
and to `routing_map` to generate them for the new route map.

EndpointRegistry memory consumption is improved too
thanks to smaller data to store and to the usage of __slots__.

OCA-git-bot committed:
endpoint_route_handler 14.0.1.0.2

simahawk committed:
Misc fix of authorship name

OCA-git-bot committed:
endpoint_route_handler 14.0.1.0.3

simahawk committed:
endpoint_route_handler: fix rules by group

OCA-git-bot committed:
endpoint_route_handler 14.0.1.0.4

simahawk committed:
endpoint_route_handler: dev status = Beta

OCA-git-bot committed:
endpoint_route_handler 14.0.1.1.0

simahawk committed:
endpoint_route_handler: move to OCA/web-api

simahawk committed:
endpoint_route_handler: fix cross worker lookup

OCA-git-bot committed:
[UPD] README.rst

Commit:
[UPD] Update endpoint_route_handler.pot

Commit:
[UPD] Update endpoint_route_handler.pot

simahawk committed:
endpoint_route_handler: add flag to control sync

To avoid multiple invalidation of all envs on each edit or create
of persistent records, a new flag is introduced: 'registry_sync'.

This flag delays the sync of the rule registry till manual action
occurs. Records in the UI are decorated accordingly to notify users of
the need to reflect changes on ther registry to make them effective.

The sync happens in a post commit hook to ensure all values are in place
for the affected records.

simahawk committed:
endpoint_route_handler: fix typo in validator

simahawk committed:
endpoint_route_handler: add constraints

simahawk committed:
endpoint_route_handler: add auto timestamp to routes

simahawk committed:
endpoint_route_handler: migrate to v16

simahawk committed:
endpoint_route_handler: add new tool model

simahawk committed:
endpoint_route_handler: adapt to new dispatch

lmignon commented on endpoint_route_handler/models/endpoint_route_sync_mixin.py:
```diff
@@ -0,0 +1,118 @@
+    )
+
+    def write(self, vals):
+        if any([x in vals for x in self._routing_impacting_fields() + ("active",)]):
```
`_routing_impacting_fields`is not defined here and should be mentioned to be implemented into consumers of this mixin

simahawk replied:
I added it on v14 and not ported the fix here. I'll do it.
Thanks for testing :)

lmignon reviewed (approved):
Thank you @simahawk Works like a charm. I refactored the code of [odoo-addon-fastapi](https://github.com/OCA/rest-framework/pull/291/commits/e510d123acdc451a3e5f7cee9ea7a0fbc347f402) to use your new mixin and everything is smooth.

lmignon commented on endpoint_route_handler/models/ir_http.py:
```diff
@@ -0,0 +1,82 @@
+
+    @classmethod
+    def routing_map(cls, key=None):
+        last_update = cls._get_routing_map_last_update(http.request.env)
```
@simahawk To be efficient, we should not rely on the last_update but on the last value of a dedicated sequence. The computation of the last_update requires a query on the whole table followed by a sort on the last_update column. This kind of operation is far more resources consuming than a simple query to get the `last_value` from a sequence.

lmignon commented on endpoint_route_handler/registry.py:
```diff
@@ -0,0 +1,342 @@
+
+            cr.execute(
+                """
+                CREATE OR REPLACE FUNCTION endpoint_route_set_timestamp()
+                    RETURNS TRIGGER AS $$
+                BEGIN
+                    NEW.updated_at = NOW();
+                    RETURN NEW;
+                END;
+                $$ LANGUAGE plpgsql;
```
Here we should create a sequence `endpoint_version` and add a trigger to auto increment the sequence on update, insert and delete.

lmignon commented on endpoint_route_handler/registry.py:
```diff
@@ -0,0 +1,342 @@
+    def make_rule(self, *a, **kw):
+        return EndpointRule(self.cr.dbname, *a, **kw)
+
+    def last_update(self):
```
should be replaced by `def last_version(self)`

```suggestion
    def last_version(self):
          self.cr.execute("SELECT last_value FROM endpoint_version")
          return self.cr.fetchone()[0]
```

simahawk replied:
I introduced a sequence on v14 here https://github.com/OCA/web-api/pull/3/commits/3ae6ceb6ef6ba83682b5d9037792cf5d3fba178d

I'm testing. I'll fwd port soon.

flachica reviewed (approved):
Functional testing with oca/rest-framework/fastapi is ok

simahawk committed:
endpoint_route_handler: use sequence as version

simahawk commented:
All changes from v14 ported, including the use of sequence instead of timestamp to detect updates :)

simahawk committed:
endpoint_route_handler: log table setup and wipe

simahawk committed:
endpoint_route_handler: fail gracefully when sync field not ready

Depending on your modules inheritance and upgrade order
when you introduce this mixin on an existing model
it might happen that
gets called before the model's table is ready
(eg: another odoo service loading the env before the upgrade happens).
Let if fail gracefully since the hook will be called again later.

simahawk committed:
endpoint_route_handler: get rid of register_hook

As routes are registered automatically in the db after sync
there's no reason to look for non registered routes at boot.
Furthermore, this is causing access conflicts on the table
when multiple instances w/ multiple workers are spawned.

lmignon commented:
@simahawk When can we expect to have this one merged? It's blocking the `fastapi` addon and I would like to publish a first release of this since Odoo 16 is out from a long time now...

simahawk commented:
> @simahawk When can we expect to have this one merged? It's blocking the `fastapi` addon and I would like to publish a first release of this since Odoo 16 is out from a long time now...

I know... In the end I'm happy I kept it on hold because there was a bug that I fixed on v14 which was breaking SF.
I'll merge 14 soon and port the latest version here.

simahawk committed:
endpoint_route_handler: fix auto_not_found param

simahawk committed:
endpoint_route_handler: 14.0.2.0.0

Commit:
[UPD] Update endpoint_route_handler.pot

OCA-git-bot committed:
[UPD] README.rst

simahawk committed:
endpoint_route_handler: update ROADMAP

simahawk commented:
@lmignon I've done a "blind" porting of the latest version from 14.0 here.
I don't have time to test but I wanted to unlock your work if possible.
Can you please take care of testing and reviewing again pls?

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
endpoint_route_handler 14.0.2.0.1

simahawk commented:
/ocabot migration endpoint_route_handler

xAdrianCif reviewed (approved):
Implementation test and code review OK

lmignon commented:
@simahawk Thank you for the last changes. At this stage I still encounter an issue when I sync the registry. I need to restart my instance to make the  route available....  It seems it only occurs for endpoints created from a xml record the first time it's activated...  It not occurs in others cases (If I add a new endpoint, our update endpoints created from xml record after restart ...)

Not a blocking issue.

simahawk commented:
> @simahawk Thank you for the last changes. At this stage I still encounter an issue when I sync the registry. I need to restart my instance to make the route available.... It seems it only occurs for endpoints created from a xml record the first time it's activated... It not occurs in others cases (If I add a new endpoint, our update endpoints created from xml record after restart ...)
> 
> Not a blocking issue.

Thanks for your feedback!
In fact, this is by design. 
IMO you should enable endpoints explicitly. I'm open for discussion of course. Feel free to open an issue :wink: 

/ocabot merge nobump
